pubtools/pulplib/_impl/client/client.py

```diff
@@ -179,6 +181,26 @@ def search_repository(self, criteria=None):
             response_f, lambda data: self._handle_page(Repository, search, data)
         )
 
+    def _do_upload_file(self, upload_id, repo_id, file_obj):
+        upload_fts = []
+
+        if isinstance(file_obj, str):
+            file_obj = open(file_obj, "rb")
+
+        checksum = hashlib.sha256()
+        size, offset = 0, 0
+        data = file_obj.read(self._CHUNK_SIZE)
+        LOG.info("Uploading %s to repo %s", file_obj, repo_id)
+        while data:
+            checksum.update(data)
+            size += len(data)
+            upload_fts.append(self._do_upload(data, upload_id, offset))
+            offset += self._CHUNK_SIZE
```

Comment:
Seems this ought to be `offset += len(data)`.

The code just a few lines above accepted that the size of data might be less than requested, but the code at this line would calculate an incorrect offset if that happened.

Per https://docs.python.org/3/library/io.html#io.RawIOBase.read

> Fewer than size bytes may be returned if the operating system call returns fewer than size bytes.

And per https://linux.die.net/man/2/read

> It is not an error if [the number of bytes read] is smaller than the number of bytes
> requested; this may happen [...] because read() was interrupted by a signal

Reply:
wow, I'd say that's very hard to catch.